utility/src/main/java/org/oppia/util/parser/CustomTagHandler.kt

```diff
@@ -0,0 +1,158 @@
+package org.oppia.util.parser
+
+import android.content.Context
+import android.text.Editable
+import android.text.Html
+import android.text.Spanned
+import android.text.style.LeadingMarginSpan
+import org.oppia.util.R
+import org.oppia.util.parser.CustomTagHandler.ListItemTag
+import org.oppia.util.parser.StringUtils.LI_TAG
+import org.oppia.util.parser.StringUtils.OL_TAG
+import org.oppia.util.parser.StringUtils.UL_TAG
+import org.xml.sax.XMLReader
+import java.util.*
+
+// TODO(#562): Add screenshot tests to check whether the drawing logic works correctly on all devices.
+/**
+ * Implements support for ordered ({@code <ol>}) and unordered ({@code <ul>}) lists in to Android TextView.
+ * <ul> and <ol> tags are added to the [listParents] Stack and removed when the closing tag is reached.
+ * <li> tags are handled [ListItemTag] instance corresponding to the parent tag.
+ *
+ * Reference: https://github.com/daphliu/android-spannable-list-sample/tree/master/app/src/main/java/com/daphneliu/sample/listspansample
+ */
+class CustomTagHandler(
+  private val context: Context
+) : Html.TagHandler {
+  private val indent = context.resources.getDimensionPixelSize(R.dimen.bullet_leading_margin)
+  private val listItemIndent = indent * 2
+  private val bulletSpan = BulletSpanWithRadius(context, indent)
+
+  private val listParents: Stack<ListItemTag> = Stack<ListItemTag>()
+
+  override fun handleTag(
+    opening: Boolean,
+    tag: String,
+    output: Editable,
+    xmlReader: XMLReader?
+  ) {
+    when (tag) {
+      UL_TAG -> {
+        if (opening) {
+          listParents.push(UnorderedListTag())
+        } else {
+          listParents.pop()
+        }
+      }
+      OL_TAG -> {
+        if (opening) {
+          listParents.push(OrderedListTag())
+        } else {
+          listParents.pop()
+        }
+      }
+      LI_TAG -> {
+        try {
+          if (opening) {
+            listParents.peek().openItem(output)
+          } else {
+            listParents.peek().closeItem(output, listParents.size)
+          }
+        } catch (e: EmptyStackException) {
+          throw EmptyStackException()
+        }
+      }
+      else -> {
+        // Found an unsupported tag $tag
+      }
+    }
+  }
+
+  /** Abstract super class for [UnorderedListTag] and [OrderedListTag]. */
+  private abstract class ListItemTag {
+    /**
+     * Opens a new list item.
```

Comment:
Here & for other methods in ``ListItemTag``, you can probably omit these KDocs if you used a sealed class since it should help self-document a bit better. We may want to add some additional documentation, but that would be to specifically clarify certain aspects of the classes.

Reply:
Done.